### PR TITLE
Add BMM150 backend

### PR DESCRIPTION
**Pull request by zehortigoza**

### CODE REVIEW

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <AP_HAL/AP_HAL.h>
+
+#include "AP_Compass_BMM150.h"
+
+#include <utility>
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Wrong include order.

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+#define DIG_Z1_LSB_REG 0x6A
+#define DIG_Z1_MSB_REG 0x6B
+#define DIG_XYZ1_LSB_REG 0x6C
+#define DIG_XYZ1_MSB_REG 0x6D
+#define DIG_Z3_LSB_REG 0x6E
+#define DIG_Z3_MSB_REG 0x6F
+#define DIG_XY2_REG 0x70
+#define DIG_XY1_REG 0x71
+
+#define MICRO_SECONDS_BETWEEN_MEASUREMENTS 10000
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: MEASURE_TIME_USEC

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    _dig_y1 = (int8_t)data[DIG_Y1_REG - DIG_X1_REG];
+    _dig_x2 = (int8_t)data[DIG_X2_REG - DIG_X1_REG];
+    _dig_y2 = (int8_t)data[DIG_Y2_REG - DIG_X1_REG];
+    _dig_xy1 = data[DIG_XY1_REG - DIG_X1_REG];
+    _dig_xy2 = (int8_t)data[DIG_XY2_REG - DIG_X1_REG];
+    _dig_z1 = (uint16_t)le16toh(data[DIG_Z1_LSB_REG - DIG_X1_REG] | data[DIG_Z1_MSB_REG - DIG_X1_REG] << 8);
+    _dig_z2 = (int16_t)le16toh(data[DIG_Z2_LSB_REG - DIG_X1_REG] | data[DIG_Z2_MSB_REG - DIG_X1_REG] << 8);
+    _dig_z3 = (int16_t)le16toh(data[DIG_Z3_LSB_REG - DIG_X1_REG] | data[DIG_Z3_MSB_REG - DIG_X1_REG] << 8);
+    _dig_z4 = (int16_t)le16toh(data[DIG_Z4_LSB_REG - DIG_X1_REG] | data[DIG_Z4_MSB_REG - DIG_X1_REG] << 8);
+    _dig_xyz1 = (uint16_t)le16toh(data[DIG_XYZ1_LSB_REG - DIG_X1_REG] | data[DIG_XYZ1_MSB_REG - DIG_X1_REG] << 8);
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Would be better to declare a packed struct here in this method instead of doing all of these `-  DIG_X1 _REG`. Then the use of le16toh() actually makes sense.

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <AP_HAL/AP_HAL.h>
+
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Missing ifdefs for Linux boards only

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    /* Do a soft reset */
+    _dev->write_register(POWER_AND_OPERATIONS_REG, SOFT_RESET);
+    hal.scheduler->delay(2);
+
+    /* Change power state from suspend mode to sleep mode */
+    _dev->write_register(POWER_AND_OPERATIONS_REG, POWER_CONTROL_VAL);
+    hal.scheduler->delay(2);
+
+    uint8_t val;
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Initialize Val to 0, so if the read_registers() call fails you don't compare the return with garbage

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+     * - Rep X/Y = 47
+     * - Rep Z = 83
+     * - ODR = 20
+     * But we are going to use 30Hz of ODR
+     */
+    _dev->write_register(REPETITIONS_XY_REG, (47 - 1) / 2);
+    _dev->write_register(REPETITIONS_Z_REG, 83 - 1);
+    /* Change operation mode from sleep to normal and set ODR */
+    _dev->write_register(OP_MODE_SELF_TEST_ODR_REG, NORMAL_MODE | ODR_30HZ);
+
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: I'm missing error checks on these calls

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+fail_sem:
+    hal.scheduler->resume_timer_procs();
+    return false;
+}
+
+/*
+ * Compensation algorithm got from https://github.com/BoschSensortec/BMM050_driver
+ * this is not explained in datasheet.
+ */
+int16_t AP_Compass_BMM150::_compensate_xy(int16_t xy, uint16_t rhall, bool is_x)
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Remove `is_x` and parametrize the function to be use for both.

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+        rhall = _dig_xyz1;
+    }
+
+    if (is_x) {
+        txy1 = _dig_x1;
+        txy2 = _dig_x2;
+    } else {
+        txy1 = _dig_y1;
+        txy2 = _dig_y2;
+    }
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: I.e.: the function receives txy1 and txy2 as parameter

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+    /* TODO: make compensation calculations works on big-endian CPUs */
+    val = ((int16_t)(((uint16_t)((((int32_t)_dig_xyz1) << 14) / rhall)) -
+          ((uint16_t)0x4000)));
+    val = ((int16_t)((((int32_t)xy) * ((((((((int32_t)_dig_xy2) *
+          ((((int32_t)val) * ((int32_t)val)) >> 7)) + (((int32_t)val) *
+          ((int32_t)(((int16_t)_dig_xy1) << 7)))) >> 9) +
+          ((int32_t)0x100000)) * ((int32_t)(((int16_t)txy2) +
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Wow. -ETOOMANYCASTS. if you reduce the number of casts, maybe it's possible to remove the TODO above

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+int16_t AP_Compass_BMM150::_compensate_z(int16_t z, uint16_t rhall)
+{
+    int32_t val;
+
+    /* TODO: make compensation calculations works on big-endian CPUs */
+    val = (((((int32_t)(z - _dig_z4)) << 15) - ((((int32_t)_dig_z3) *
+          ((int32_t)(((int16_t)rhall) - ((int16_t)_dig_xyz1)))) >> 2)) /
+          (_dig_z2 + ((int16_t)(((((int32_t)_dig_z1) *
+          ((((int16_t)rhall) << 1))) + (1 << 15)) >> 16))));
+
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Ditto

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+    if (time_us - _last_update_timestamp < MICRO_SECONDS_BETWEEN_MEASUREMENTS) {
+        return;
+    }
+
+    if (!_dev->get_semaphore()->take_nonblocking()) {
+        return;
+    }
+
+    uint16_t data[4];
+    _dev->read_registers(DATA_X_LSB_REG, (uint8_t *) &data, sizeof(data));
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Missing error check here. Otherwise data will contain garbage from stack

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    val = _compensate_xy(val, rhall, true);
+    raw_field.x = val;
+
+    val = le16toh(((int16_t)data[1]) >> 3);
+    val = _compensate_xy(val, rhall, false);
+    raw_field.y = val;
+
+    val = le16toh(((int16_t)data[2]) >> 1);
+    val = _compensate_z(val, rhall);
+    raw_field.z = val;
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: These shift right look suspicious. Are you sure about them?

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+
+    /* apply sensitivity scale */
+    raw_field /= 16;
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Should be a define so if we ever change the scale we don't forget this here

File: libraries/AP_Compass/AP_Compass_BMM150.cpp [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+void AP_Compass_BMM150::read()
+{
+    if (_accum_count == 0) {
+        return;
+    }
+
+    hal.scheduler->suspend_timer_procs();
+
+    Vector3f field(_mag_accum);
```
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: _mag_acumm is changed above without any lock held.

File: libraries/AP_Compass/AP_Compass_BMM150.cpp (hunk 4 above)
Comment (zehortigoza) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Why only build for Linux boards?
The build for px4 failed for this PR but is because it did not found the definition of le16toh.

Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: it's currently not supported since it doesn't implement I2CDevice and friends

Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: and... since it's not tested there, shouldn't be enabled.

For le16toh() you need to (unconditionally) include sparse-endian.h. And in the struct I told you to add below, you use le16_t type where it makes sense.

File: libraries/AP_Compass/AP_Compass_BMM150.cpp (hunk 12 above)
Comment (zehortigoza) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: The register 0x48 bit field is:
BIT0 = data ready status
BIT1 = fixed 0
BIT2 = RHALL LSB BIT0
BIT3 = RHALL LSB BIT1
...

Register 0x42 bit field:
BIT0 = X-Self test
BIT1 = fixed 0
BIT2 = fixed 0
BIT3 = RAW X LSB BIT0
BIT4 = RAW X LSB BIT1
...
Similar to what happen to raw Y and raw Z, that is why this shifts are done.

File: libraries/AP_Compass/AP_Compass_BMM150.cpp (hunk 13 above)
Comment (zehortigoza) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: We can't change the scale on this sensor, it will always be 16 LSB/uT.

File: libraries/AP_Compass/AP_Compass_BMM150.cpp (hunk 14 above)
Comment (zehortigoza) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Huum I was thinking that read() and also run in timer thread but the read() runs on the main thread.
All other Compas also don't held a lock, so should I send another commits locking?

Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: It depends where they release the global timer lock

Comment (zehortigoza) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Looking again I still can't see where they hold a lock in the _update() function.

Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Compass/AP_Compass_AK8963.cpp#L281

suspend_timer_proc() has a weird name, but it's actually a "global lock" protecting data changes between timer and main threads. It's not ideal IMO, but it's the way it's done for now.

Comment (zehortigoza) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: Ha? on the line 281 of AK8963 it is just releasing the lock of the bus.
And there is no call to suspend_timer_proc() on the _update() of AK8963.

Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_BMM150.cpp]: I apparently misread your code. The _update method runs with the lock taken (it's actually taken by the caller). I.e. your code is correct in this part.

**Comment by zehortigoza** [comment]
New version here https://github.com/ArduPilot/ardupilot/pull/4422
